### DISCUSSION TITLE
Why a field marked as @hybrid_property receives a InstrumentedAttribute?

### DISCUSSION THREAD

**Discussion by uselessscat** [sqlalchemy/sqlalchemy#6534] [Usage Questions] [answered]
Hi,

I have this definition:

```python
import re
from sqlalchemy.ext.declarative import as_declarative
from sqlalchemy.ext.declarative import DeferredReflection
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy import event, Table

@as_declarative()
class Model:
    __abstract__ = True

class MyTable(DeferredReflection, Model):
    __tablename__ = 'my_table'

    @event.listens_for(Table, 'column_reflect')
    def rename_reflective_columns(inspector, table, column):
        name = column['name']

        if name in {'my_field'}:
            column['key'] = f'_{name}'

    @hybrid_property
    def my_field(self):
        r = r'[a-z]*' # Just an example

        matches = re.findall(r, self._my_field)
        valid = []

        for match in matches:
            valid.append(match)

        return valid
        
    @my_field.setter
    def my_field(self, f):
        self._my_field = f
```

So im trying to query the validated field in a test case but im having this error:

```
//// Hidden stacktrace -- model(*args, **kwargs) is the same as MyTable(*args, **kwargs)
tests/models/fixtures.py:19: in _add_model
    new = model(*args, **kwargs)
<string>:4: in __init__
    ???
../local/lib/python3.8/site-packages/sqlalchemy/orm/state.py:433: in _initialize_instance
    manager.dispatch.init_failure(self, args, kwargs)
../local/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:68: in __exit__
    compat.raise_(
../local/lib/python3.8/site-packages/sqlalchemy/util/compat.py:182: in raise_
    raise exception
../local/lib/python3.8/site-packages/sqlalchemy/orm/state.py:430: in _initialize_instance
    return manager.original_init(*mixed[1:], **kwargs)
../local/lib/python3.8/site-packages/sqlalchemy/ext/declarative/base.py:838: in _declarative_constructor
    if not hasattr(cls_, k):
../local/lib/python3.8/site-packages/sqlalchemy/ext/hybrid.py:898: in __get__
    return self._expr_comparator(owner)
../local/lib/python3.8/site-packages/sqlalchemy/ext/hybrid.py:1105: in expr_comparator
    comparator(owner),
../local/lib/python3.8/site-packages/sqlalchemy/ext/hybrid.py:1090: in _expr
    return ExprComparator(cls, expr(cls), self)
app/auth/model.py:128: in my_field
    matches = re.findall(r, self._my_field)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pattern = '[a-z]*', string = <sqlalchemy.orm.attributes.InstrumentedAttribute object at 0x7f49a0f4cd60>, flags = 0

    def findall(pattern, string, flags=0):
>       return _compile(pattern, flags).findall(string)
E       TypeError: expected string or bytes-like object

../local/lib/python3.8/re.py:241: TypeError
```

So I wonder why this is receiving InstrumentedAttibute instead of the model instance?, and how can i fix that.

Thanks

**Comment by zzzeek** [comment, marked as answer]
hi -

you cant use this hybrid as written in a query because it does not provide valid [@expression](https://docs.sqlalchemy.org/en/14/orm/extensions/hybrid.html#defining-expression-behavior-distinct-from-attribute-behavior) behavior.    When used as MyTable.my_field, "self" is actually the MyTable class, and "self._my_field" we would assume is a Column of some kind but it's not shown here; but in any case a function like `re.findall()` applies to Python strings, not to SQLAlchemy expression constructs.   

To better understand the difference, pretend there's no "hybrid" here and there is just the MyTable class.  What SQL are you looking to send to the database when you write this query?      Can you write the SQL you are looking for if you didn't have a MyTable class or the SQLAlchemy ORM?  you likely need to use a SQL function like [match](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.ColumnOperators.match) although it seems you are returning a list here so you might need to consult your database documentation for SQL functions you are looking for here.

**Reply by uselessscat** [reply]
Hi @zzzeek, I dont want to modify the way the query is built (for now), Just the data I will get from that field. Also I need to use the field in filters (for now as a varchar column).

Here is more context. The field `_my_field` is defined as VARCHAR but because of unknown reason the older systems save a list inside this column. I cant change this soon because is used in many places i dont know so i have to parse as is and return valid values to my service. The content of the column is the same as `str(['value1', 'value2', 'etc'])` so i want to parse it with regexes and match them between a predefined set (ENUM) in the new system. 

This is why I was using @hybrid_field, but as I understand now I will get the value of the object and the column object when I user this decorator. I used to think that i will just receive the value but not. So I can reframe the question to :

How can I only get the value and bypass/dont do anything with the query (for now) ? 

I tested a condition like 

```python
if isinstance(self._my_field, InstrumentedAttribute):
    # do stuff
    return correct_data
else:
    return self._my_field
```

But im not sure if this is correct and always will work as I expected, can you help me with this?

**Reply by zzzeek** [reply]
> Hi @zzzeek, I dont want to modify the way the query is built (for now), 

forgive me if this sounds blunt but AFIACT there is no working query right now, you are getting an error message because you are attempting to use Python's re.findall as a SQL construct, which is not possible.  it sounds like you don't actually intend to use `@hybrid_property` and should be using `@property` instead, if you'd like this data to be converted from a list in Python after results have been received from the database.


> Here is more context. The field `_my_field` is defined as VARCHAR but because of unknown reason the older systems save a list inside this column. I cant change this soon because is used in many places i dont know so i have to parse as is and return valid values to my service. The content of the column is the same as `str(['value1', 'value2', 'etc'])` so i want to parse it with regexes and match them between a predefined set (ENUM) in the new system.

SQL databases do not have the same regexp capabilities as that of Python.    I would advise not using "my_field" in a SQL query and instead parse out the data once you have the object.  That is, replace the use of `@hybrid_property` with `@property` and then select "_my_field" in your SQL query instead.



> 
> This is why I was using @hybrid_field, but as I understand now I will get the value of the object and the column object when I user this decorator. I used to think that i will just receive the value but not. So I can reframe the question to :
> 
> How can I only get the value and bypass/dont do anything with the query (for now) ?

query for MyTable._my_field, i.e. the Column object directly, rather than trying to use a hybrid.

**Reply by uselessscat** [reply]
I understand, good point. I will change then

Thanks
